distributed/shuffle/tests/test_no_pandas_import.py

```diff
@@ -0,0 +1,32 @@
+from __future__ import annotations
+
+import subprocess
+import sys
+
+import_check_code = """
+import sys
+
+current_pandas_modules = [m for m in sys.modules if m.startswith("pandas")]
+assert (
+    not current_pandas_modules
+), "pandas is already imported at startup: " + "\\n".join(current_pandas_modules)
+
+# Make pandas un-importable
+sys.modules["pandas"] = None
+# "if the value is None, then a ModuleNotFoundError is raised"
+# https://docs.python.org/3.6/reference/import.html#the-module-cache
+
+import distributed.shuffle  # noqa: F401
+"""
+
+
+def test_import_no_pandas():
+    p = subprocess.run(
+        [sys.executable],
+        input=import_check_code,
+        text=True,
+        timeout=10,
+    )
+    assert (
+        p.returncode == 0
+    ), "Importing the shuffle extension without pandas failed. See logs for details."
```

Comment:
```suggestion
import sys

from distributed.nanny import Nanny
from distributed.utils_test import gen_cluster


@gen_cluster(client=True, Worker=Nanny)
async def test_import_no_pandas(c, s, a, b):
    def assert_no_pandas(dask_worker):
        assert dask_worker.extensions["shuffle"]
        assert "pandas" not in sys.modules
        
    await c.run(assert_no_pandas)
```

Reply:
This is a neat way of doing it. The one thing is that it won't point you to _where_ pandas is being imported, just that it is (somewhere). The reason I went for running the script in a subprocess was that you'd get an import error (and traceback to) exactly where the import was happening, which made fixing the failure much easier.

If you'd prefer giving up the debuggability for this more conventional pattern, I'm happy to make the change.

Reply:
I think that git bisect is enough to offer debuggability?

Reply:
Even still, you have to figure out what line of code is causing the import (and know how to use git bisect). Getting the traceback to the exact line of code just feels more friendly to other contributors.

Reply:
I counter-argue that your version would not notice it if there was an ``import pandas`` inside ``ShuffleWorkerExtension.__init__``. IMHO tests should be as black-box as possible.

Reply:
That's a good point.

Unfortunately, the `gen_cluster` approach doesn't work. Pandas is already imported even if the shuffle extension is not imported on the worker. A good example of how when there isn't a traceback to see where the import is happening, debugging it is tough!

My guess is that the problem is that nannies pre-import pandas (and other modules) automatically into the worker process:
The global `mp_context` used by Nannies to create worker processes is initialized here at import time: https://github.com/dask/distributed/blob/338d0be2e50b5458ef019ec8ce7377bda3097ef1/distributed/utils.py#L69-L93
In particular, notice how all required and optional packages that are importable are added to `set_forkserver_preload`.

pandas is in the list of optional packages: https://github.com/dask/distributed/blob/338d0be2e50b5458ef019ec8ce7377bda3097ef1/distributed/versions.py#L24-L29

(Sidenote: we're using [`spawn` by default](https://github.com/gjoseph92/distributed/blob/338d0be2e50b5458ef019ec8ce7377bda3097ef1/distributed/distributed.yaml#L77); I confirmed that's what's used when I run the test. Reading the [`multiprocessing` code](https://github.com/python/cpython/blob/dad6be5ffe48beb74fad78cf758b886afddc7aed/Lib/multiprocessing/context.py#L180-L185), I don't understand why `set_forkserver_preload` is affecting non-forkserver contexts. But if I comment out the pandas line from `optional_packages`, the test passes, so I'm not going to look into it much further.)

Reply:
For posterity: @crusaderky and I discovered that `set_forkserver_preload` has no effect when using `spawn`, as expected. Instead, the reason pandas was always imported on Nanny workers (but not imported if it was commented out from the list in `versions.py`) is that the `get_versions` check on client connect imports all those packages listed in order to get their versions. This could be avoided.